routing/src/prefix.rs

```diff
@@ -142,6 +142,17 @@ impl From<Ipv6Prefix> for Prefix {
         Self::IPV6(value)
     }
 }
+impl From<&str> for Prefix {
```

Comment:
can we either

1. add `#[cfg(test)]` to this or
2. put this in a `tests` module?

I understand wanting this in tests, but panic is not an ok answer to invalid input on a day to day basis and we should not allow this type of logic into the rest of the dataplane.

Reply:
I'm fine with making them test-only, this was my use case anyway. There are some other surrounding implementations of the `From` trait that can `panic` as well, I can move them at the same time if it doesn't break anything.

Reply:
(Assuming I come back to this PR, not sure based on what Fredi said.)

Reply:
That's fair.

Generally, I don't like impls of `From` or `Into` which can panic.

Another option is to use a macro for situations like this

Reply:
How does the macro prevent the panic? You do have to parse the string into a valid prefix at one point or another?

Reply:
I agree with @daniel-noland with panicking in prod code. I've pushed 3 commits to the management cleanup PR (which should be merged soon) that address a very similar thing. Initially the routing crate had a testing-only converting function (annotated with `#[cfg(test)]`. Later on, the annotation was dropped (I believe by me) and as such it was used by gRPC prod code which was panicking when the input configuration contained garbage prefixes. The first commit is https://github.com/githedgehog/dataplane/pull/441/commits/4f939261055330b7fae04502572ae2de0453465b

I've annotated it back with `#[cfg(test)]`. However, to allow using it from other crates, I've had to define a feature flag (`'testing'`) in the routing crate and enable it wherever it is needed.

Reply:
> How does the macro prevent the panic? You do have to parse the string into a valid prefix at one point or another?

The `#[cfg(test)]` macro means that the panic implementation simply does not exist in production code.  It is fine to panic in the test :)

Reply:
> How does the macro prevent the panic? You do have to parse the string into a valid prefix at one point or another?

The `#[cfg(test)]` macro means that the panic implementation simply does not exist in production code.  It is fine to panic in the test :)

Reply:
Aaah OK I didn't understand what you suggested, I thought you meant to implement the trait or function as a macro somehow. Got it now :slightly_smiling_face: